specs/TransportProtocols.md

```diff
@@ -0,0 +1,176 @@
+# Transport Protocols
+
+This document describes the protocols used by the three ASP.NET Endpoint Transports: WebSockets, Server-Sent Events, Long Polling and HTTP Post
+
+## Transport Requirements
+
+A transport is required to have the following attributes:
+
+1. Duplex - Able to send messages from Server to Client and from Client to Server
+1. Frame-based - Messages have fixed length (as opposed to streaming, where data is just pushed throughout the connection)
+1. Frame Metadata - Able to encode two pieces of Frame Metadata
+    * `Type` - Either `Binary` or `Text`.
+    * `EndOfMessage` - Boolean indicating if this frame represents the final frame of a complete message.
+1. Binary-safe - Able to transmit arbitrary binary data, regardless of content
+1. Text-safe - Able to transmit arbitrary text data, preserving the content. Line-endings must be preserved **but may be converted to a different format**. For example `\r\n` may be converted to `\n`. This is due to quirks in some transports (Server Sent Events)
+
+The only transport which fully implements the duplex requirement is WebSockets, the others are "half-transports" which implement one end of the duplex connection. They are used in combination to achieve a duplex connection.
+
+Throughout this document, the term `[endpoint-base]` is used to refer to the route assigned to a particular end point
+
+## WebSockets (Full Duplex)
+
+The WebSockets transport is unique in that it is full duplex, and a persistent connection that can be established in a single operation. As a result, the client is not required to use the `[endpoint-base]/negotiate` endpoint to establish a connection in advance. It also includes all the necessary metadata in it's own frame metadata.
+
+The WebSocket transport is activated by making a WebSocket connection to `[endpoint-base]/ws`. Upon doing so, the connection is fully established and immediately ready for frames to be sent/received. The WebSocket OpCode field is used to indicate the type of the frame (Text or Binary) and the WebSocket "FIN" flag is used to indicate the end of a message.
+
+Establishing a second WebSocket connection when there is already a WebSocket connection associated with the Endpoints connection is not permitted and will fail with a `400` (Bad Request) status code.
+
+## HTTP Post (Client-to-Server only)
```

Comment:
The big problem, it seems, is that there isn't really a process to allow the server to terminate the open request gracefully. Only the client can gracefully terminate the connection because they need to write a chunk terminator. We could try just sending the response while the client is still writing chunks, but I'm not confident that would trigger the desired behavior. Obviously we can disgracefully terminate, but then it's very difficult for the client to know how much the server has read.

We can do some testing an examining, but in the mean time, I'm kinda liking the idea of the `/send` endpoint using the same protocol as the long-polling responses. For now, let's just specify that the server is not expected to deliver any of the messages until the entire response is received (i.e. streaming messages won't work, yet).